Time stepper interface adjustments

wthrowe opened the pull request:
## Proposed changes

* Adds dedicated classes for storing time stepper history
* Restricts boundary calculations to two sides

### Types of changes:

- [ ] Bugfix
- [X] New feature

### Component:

- [X] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`


### Further comments

The volume history class (History) does not have many features beyond a deque and could be dropped if people prefer.  I added it mostly for consistency with the boundary interface and as a possible place to do allocation caching if we end up wanting that.

wthrowe committed:
Generate long AB sums with metaprogramming

wthrowe committed:
Add pup function for std::deque

wthrowe committed:
Add class for holding TimeStepper history

kidder commented on src/Time/BoundaryHistory.hpp:
```diff
@@ -0,0 +1,310 @@
+
+ private:
+  template <size_t Side>
+  void clean_cache(const Time* const goner) noexcept;
```
I'd prefer a better name than `goner` as it is a bit informal/slang, perhaps `to_remove` or `to_be_removed`

kidder commented on src/Time/BoundaryHistory.hpp:
```diff
@@ -0,0 +1,310 @@
+  const auto insert_result = coupling_cache_.insert(
+      std::make_pair(std::make_pair(&*local, &*remote), BoundaryVars{}));
+  BoundaryVars& inserted_value = insert_result.first->second;
+  const bool new_value = insert_result.second;
```
new_value -> is_new_value

kidder commented on src/Time/BoundaryHistory.hpp:
```diff
@@ -0,0 +1,310 @@
+  const bool new_value = insert_result.second;
+  if (new_value) {
+    inserted_value =
+        std::forward<Coupling>(c)(as_const(std::get<2>(*local.base_)),
```
There is a c++17 as_const; we should probably move ours to the cpp17 namespace

kidder commented on src/Time/BoundaryHistory.hpp:
```diff
@@ -0,0 +1,321 @@
+
+ private:
+  template <size_t Side>
+  void clean_cache(const Time* goner) noexcept;
```
I'd prefer a better name than goner as it is a bit informal/slang, perhaps to_remove or to_be_removed (I'm repeating this as it seems to have disappeared as outdated)

kidder commented on src/Time/History.hpp:
```diff
@@ -0,0 +1,192 @@
+  /// iterators using HistoryIterator::value() and
+  /// HistoryIterator::derivative().
+  //@{
+  const_iterator begin() const noexcept { return data_.begin(); }
```
I am confused by the comment; these return `HistoryIterator`s (which when dereferenced give the Time)

wthrowe replied:
Yeah, it's poorly worded.  I want some comment to this effect in the `History` documentation because I'd look for it there and not in the iterator documentation, and having no explanation in `History` of how to get values out sounds confusing.  (I should add a similar thing to `BoundaryHistory`.)  I'll improve it.

kidder commented on src/Time/History.hpp:
```diff
@@ -0,0 +1,192 @@
+  const_iterator cend() const noexcept { return end(); }
+
+  size_type size() const noexcept { return data_.size(); }
+  const_reference operator[](size_type n) const noexcept {
```
so I would move the comment about getting directly the Time to here...

kidder commented on src/Time/History.hpp:
```diff
@@ -0,0 +1,192 @@
+
+  HistoryIterator() = default;
+
+  reference operator*() const noexcept { return std::get<0>(*base_); }
```
and I would add something here about these giving the Time

wthrowe committed:
Move as_const to cpp17 namespace

wthrowe committed:
Reuse allocations in History

wthrowe commented:
I've been looking at how the boundary calculations should be done for non-Adams-Bashforth integrators and I think that the correct boundary "value" stored in the BoundaryHistory may always be zero.  If so that would simplify this code a fair amount.

I need to go through the math more carefully to check this.  I don't have a strong preference about whether we merge this and then fix it later if necessary or delay merging until I've done any needed changes.

wthrowe commented:
I think I'm going to split this and do the volume parts here and save the boundary parts for another pull request.

wthrowe committed:
Use History in time steppers

kidder commented:
okay sounds good

wthrowe commented:
I've removed the boundary changes other than moving the history clearing into the TimeSteppers.  As a bonus I've added code to reuse old history allocations for the next step's time derivatives.

wthrowe commented:
There are two function declarations for removed functions remaining in RungeKutta3.hpp.  This is a reminder to myself to remove them.

kidder commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -174,6 +174,24 @@ constexpr T min_by_magnitude(std::initializer_list<T> ilist) {
+/// \ingroup ConstantExpressionsGroup
+/// \brief Returns `f(ic<0>{}) + f(ic<1>{}) + ... + f(ic<NumTerms-1>{})`
+/// where `ic<N>` stands for `std::integral_constant<size_t, N>`.
+template <size_t NumTerms, typename Function,
```
comment why this is not a for loop

kidder commented on src/Time/History.hpp:
```diff
@@ -0,0 +1,196 @@
+/// History data used by a TimeStepper.
+/// \tparam Vars type of variables being integrated
+/// \tparam DerivVars type of derivative variables
+template <typename Vars, typename DerivVars>
```
can we default DerivVars?

wthrowe replied:
Turns out we can but doing so is not useful because this is almost always used in generic code that is already templated on Vars and DerivVars and passes them through to this class.  The only place where a default could be used is in one section of tests, and the required default would be `DerivVars=Vars` which is not the norm for evolutions.

kidder commented on src/Time/History.hpp:
```diff
@@ -0,0 +1,196 @@
+// ================================================================
+
+template <typename Vars, typename DerivVars>
+void History<Vars, DerivVars>::insert(Time time, Vars value,
```
do you want to inline these functions

kidder commented on tests/Unit/Time/Test_History.cpp:
```diff
@@ -0,0 +1,131 @@
+  CHECK(next_it + (-1) == it);
+  CHECK(0 + it == it);
+  CHECK(1 + it == next_it);
+  CHECK((-1)+ next_it == it);
```
formatting

kidder commented on src/Time/History.hpp:
```diff
@@ -0,0 +1,196 @@
+
+  /// Add a new set of values at the start of the history.  This is
+  /// often convenient for setting initial data.
+  void insert_initial(Time time, Vars value, DerivVars deriv) noexcept;
```
note this adds the value to the front

kidder reviewed (changes requested):
fix and squash
